src/persist-client/src/internal/trace.rs

```diff
@@ -207,10 +205,14 @@ impl<T: Timestamp + Lattice> Trace<T> {
             let desc = batch.desc.clone();
             let mut parts = Vec::with_capacity(batch.parts.len());
             let mut descs = Vec::with_capacity(batch.parts.len());
-            for id_batch in &batch.parts {
-                parts.push(id_batch.id);
-                descs.push(id_batch.batch.desc.clone());
-                legacy_batches.insert(Arc::clone(&id_batch.batch), ());
+            for IdHollowBatch { id, batch } in &batch.parts {
+                parts.push(*id);
+                descs.push(batch.desc.clone());
+                if batch.desc.lower() == batch.desc.upper() {
```

Comment:
This is the "one weird trick" of this PR.

IIUC, when we tried the empty-time-range-batch change in the past, we ran into weird errors due to how hollow batches were no longer unique: calling `exert` twice in a row without changing the frontiers would insert two identical batches into the spine. The `legacy_batches` collections in spine state / diffs are kind-of differential collections, but they're not built to handle cardinalities > 1.

We have a solution lying around for this - the `hollow_batches` collection, where batches are unique by virtue of having a unique ID. We didn't migrate old hollow batches into this collection originally, since we were worried about generating huge diffs before we had bounded state sizes. However, there's no risk involved in putting our new empty-time-range batches there, and it solves the ambiguity problem.

An alternative would be to track which collection each batch ought to go in at the batch level, and start putting all new batches in the new location. More invasive but puts us farther along in the migration... though we'd still be juggling both collections indefinitely.

Reply:
Cool! Probably worth sticking at least parts of this comment in the code?

Also, there's an assert below that hollow_batches is empty when we're not roundtripping structure. Time to remove the dyncfg (doesn't need to be part of this PR)?

```
        if !self.roundtrip_structure {
            assert!(hollow_batches.is_empty());
            spine_batches.clear();
            merges.clear();
        }
```

Reply:
Totally! Added the comment.

I've been hoping to remove `roundtrip_structure`, but it appears there's a very small number of live rollups that don't have structure metadata at all still... I think just shards that get almost no traffic of any sort, but I need to look into it properly. In the meantime I've noted your comment on my draft PR to make sure I don't lose track of it.